Рефакторинг тестов и обработки ошибок

stewkk opened the pull request:
- Теперь апи возвращает описание ошибки в формате `problem+json`
- К ошибкам добавлен контекст (используется стандартный пакет `errors`)
- Убрана зависимость от либы `check`
- В юнит-тестах теперь применяются [**table-driven tests**](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests)

stewkk committed:
Рефакторинг тестов для /ping

stewkk committed:
Тестирование API вынесено в отдельный пакет

stewkk committed:
Переписаны тесты для /users

stewkk committed:
Стилевые правки

stewkk committed:
Убран TODO теста для логгирования запросов

stewkk committed:
Добавлены чекеры для юнит-тестов

stewkk committed:
Рефакторинг тестов для репозитория users

Mazdaywik commented:
Извините, а что такое «рефакторинг ошибок»?

stewkk commented:
> Извините, а что такое «рефакторинг ошибок»?

Поправил, спасибо =)

stewkk committed:
Рефакторинг тестов для бизнес-логики users

stewkk committed:
Исправлен тест валидации почты

stewkk committed:
Добавлено описание ошибок api

stewkk committed:
Рефакторинг валидации и обработки ошибок

geogreck commented:
Александр, добрый день! 
Спешу вам сообщить, что у вас отвалился линтер 
А ещё у вас неправильное время на компьютере и вы пушите в будущее

stewkk committed:
Изменены хендлеры для NotFound и MethodNotAllowed

Стандартные хендлеры из httprouter заменены на кастомные. Теперь они
также возвращают описание ошибки в формате problem+json.

geogreck commented:
Отличная работа, Александр! Мне очень нравится

uma-op commented on backend/internal/errors/middleware.go:
```diff
@@ -0,0 +1,123 @@
+			if err != nil {
+				logger.Info(err)
+				w.WriteHeader(http.StatusInternalServerError)
+				return
```
Чуть ниже уже есть return

uma-op reviewed (dismissed):
Норм
